spec/amber/router/params_spec.cr

```diff
@@ -0,0 +1,81 @@
+require "../../../spec_helper"
+
+module Amber::Router
+  describe Parser do
+    headers = HTTP::Headers.new
+
+    context "when parsing query string" do
+      it "returns array of values for same param key" do
+        request = HTTP::Request.new("GET", "/?test=test&test=test2")
+
+        request.params["test"].should eq "test"
+        request.params.fetch_all("test").size.should eq 2
+        request.params.fetch_all("test").should eq %w(test test2)
+      end
+    end
+
+    context "when parsing request body" do
+      context "when content-type is form url urlencoded" do
+        it "returns params" do
+          headers["Content-Type"] = "application/x-www-form-urlencoded"
+          body = "name=John Doe"
+          request = HTTP::Request.new("POST", "/", headers, body)
+          request.params["name"].should eq "John Doe"
+        end
+
+        it "parses body params with charset" do
+          headers["Content-Type"] = "application/x-www-form-urlencoded; charset=UTF-8"
+          body = "\x68\x65\x6C\x6C\x6F\x3D\x77\x6F\x72\x6C\x64"
+          request = HTTP::Request.new("POST", "/", headers, body)
+          request.params["hello"].should eq "world"
+        end
+      end
+
+      context "when parsing multipart" do
+        it "parses files from multipart forms" do
+          headers["Content-Type"] = "multipart/form-data; boundary=fhhRFLCazlkA0dX"
+          body = "--fhhRFLCazlkA0dX\r\nContent-Disposition: form-data; name=\"_csrf\"\r\n\r\nPcCFp4oKJ1g-hZ-P7-phg0alC51pz7Pl12r0ZOncgxI\r\n--fhhRFLCazlkA0dX\r\nContent-Disposition: form-data; name=\"title\"\r\n\r\ntitle field\r\n--fhhRFLCazlkA0dX\r\nContent-Disposition: form-data; name=\"picture\"; filename=\"index.html\"\r\nContent-Type: text/html\r\n\r\n<head></head><body>Hello World!</body>\r\n\r\n--fhhRFLCazlkA0dX\r\nContent-Disposition: form-data; name=\"content\"\r\n\r\nseriously\r\n--fhhRFLCazlkA0dX--"
```

Comment:
I think a heredoc would make this easier to read, even if the `\r` is preserved:

```
--fhhRFLCazlkA0dX\r
Content-Disposition: form-data; name="_csrf"\r
\r
PcCFp4oKJ1g-hZ-P7-phg0alC51pz7Pl12r0ZOncgxI\r
--fhhRFLCazlkA0dX\r
Content-Disposition: form-data; name="title"\r
\r
title field\r
--fhhRFLCazlkA0dX\r
Content-Disposition: form-data; name="picture"; filename="index.html"\r
Content-Type: text/html\r
\r
<head></head><body>Hello World!</body>\r
\r
--fhhRFLCazlkA0dX\r
Content-Disposition: form-data; name="content"\r
\r
seriously\r
--fhhRFLCazlkA0dX--
```

Related, The spec seems to indicate that it's not proper but the HTTP body doesn't _need_ the `\r`. Do we handle parsing the raw http, or is Crystal? Do we need a test around parsing a request body with only `\n`?

Reply:
In HTTP blocks are separated by `\r\n\r\n`